pkg/client/client.go

```diff
@@ -0,0 +1,410 @@
+					c.obsLock.RUnlock()
+
+					sendMessage := func() {
+						err := checkinClient.Send(&proto.StateObserved{
+							Token:          c.token,
+							ConfigStateIdx: cfgIdx,
+							Status:         observed,
+							Message:        observedMsg,
+						})
+						if err != nil {
+							c.impl.OnError(err)
+							return
+						}
+						lastSent = time.Now()
+						lastSentCfgIdx = cfgIdx
+						lastSentStatus = observed
+						lastSentMessage = observedMsg
+					}
+
+					// On start keep trying to send the initial check-in.
+					if lastSent.IsZero() {
+						sendMessage()
+						continue
+					}
+
+					// Send new status when it has changed.
+					if lastSentCfgIdx != cfgIdx || lastSentStatus != observed || lastSentMessage != observedMsg {
+						sendMessage()
+						continue
+					}
+
+					// Send when more than 30 seconds has passed without any status change.
+					if time.Now().Sub(lastSent) >= c.minCheckTimeout {
+						sendMessage()
+						continue
+					}
+				}
+			}()
+
+			// wait for both send and recv go routines to stop before
+			// starting a new stream.
+			wg.Wait()
+		}
+	}()
+}
+
+// startActions starts the go rountines to send and receive actions
+func (c *Client) startActions() {
+	c.wg.Add(1)
+
+	// results our held outside of the retry loop, because on re-connect
+	// we still want to send the responses that either failed or haven't been
+	// sent back to the agent.
+	actionResults := make(chan *proto.ActionResponse, 100)
+	go func() {
+		defer c.wg.Done()
+		for {
+			select {
+			case <-c.ctx.Done():
+				// stopped
+				return
+			default:
+			}
+
+			actionsClient, err := c.client.Actions(c.ctx)
+			if err != nil {
+				c.impl.OnError(err)
+				continue
+			}
+
+			// initial connection of stream must send the token so
+			// the Elastic Agent knows this clients token.
+			actionResults <- &proto.ActionResponse{
+				Token:  c.token,
+				Id:     "init",
+				Status: proto.ActionResponse_SUCCESS,
+				Result: []byte("{}"),
+			}
+
+			var wg sync.WaitGroup
+			wg.Add(2)
+			done := make(chan bool)
+
+			// action requests
+			go func() {
+				defer wg.Done()
+				for {
+					action, err := actionsClient.Recv()
+					if err != nil {
+						if err != io.EOF {
+							c.impl.OnError(err)
+						}
+						close(done)
+						return
+					}
+
+					actionImpl, ok := c.actions[action.Name]
+					if !ok {
+						actionResults <- &proto.ActionResponse{
+							Token:  c.token,
+							Id:     action.Id,
+							Status: proto.ActionResponse_FAILED,
+							Result: ActionErrUndefined,
+						}
+						continue
+					}
+
+					var params map[string]interface{}
+					err = json.Unmarshal(action.Params, &params)
+					if err != nil {
+						actionResults <- &proto.ActionResponse{
+							Token:  c.token,
+							Id:     action.Id,
+							Status: proto.ActionResponse_FAILED,
+							Result: ActionErrUnmarshableParams,
+						}
+						continue
+					}
+
+					// perform the action
+					go func() {
```

Comment:
as ^^ should we make this sychronous? what if actions of conflicting nature are performed at the same time?

Reply:
Answered this above as well.